doc/manual/contracts.md

```diff
@@ -0,0 +1,635 @@
+different in nature: in the first one, the function `add_semi` respects its
+contract. Whenever `x` is a string, `add_semi` does return a string. Here, the
+**caller** (the user of the function) is to blame, and not the **callee** (the
+function). This is an important distinction. Say you wrote `add_semi` as a part
+of a library. A wrong call is a random user, somewhere, making a programming
+mistake. This is not your direct responsibility, assuming the contract error
+messages are good enough to let users understand the issue immediately.
+
+The second example is reversed. The caller provides a string, as requested by
+the contract, but the function returns a number. The blame is on the **callee**.
+If you are the library writer who shipped the `wrong` function, this means you
+have a bug that you need to fix.
+
+The interpreter automatically performs book-keeping for functions contracts in
+order to make this caller/callee distinction:
+
+```
+nickel>let add_semi | Str -> Str = fun x => x ++ ";" in
+add_semi 1
+error: contract broken by the caller.
+[..]
+
+nickel>let wrong | Str -> Str = fun x => 0 in
+wrong "a"
+error: contract broken by a function.
+[..]
+```
+
+##### Higher-order functions
+
+The beauty of function contracts is that they gracefully scale to higher-order
+functions as well. Higher-order functions are functions that take other
+functions as parameters. Here is an example:
+
+```
+nickel>let apply_fun | (Num -> Num) -> Num = fun f => f 0 in
+apply_fun (fun x => "a")
+error: contract broken by the caller.
+  ┌─ :1:9
+  │
+1 │ (Num -> Num) -> Num
+  │         --- expected return type of a function provided by the caller
+  │
+  ┌─ repl-input-17:2:21
+  │
+2 │ apply_fun (fun x => "a")
+  │                     --- evaluated to this expression
+[..]
+```
+
+#### Dictionary
+
+The type constructor `{_ : Contract}` represents a record whose field names are not
+constrained but whose field values are checked against `Contract`. In practice, the
+contract `Contract` is mapped onto each field. Such a contract is useful when using
+records as an extensible dictionary, that is a key/value store, where keys are
+strings and values respect `Contract`. Example:
+
+```nickel
+let occurrences | {_: Num} = {a = 2, b = 3, "!" = 5, "^" = 1} in
+occurrences."!"
+```
+
+## Laziness
+
+In the [writing a custom contract by hand](#by-hand) section, we noted the
+strange fact that a custom contract must return a value, instead of just
+returning e.g. a boolean to indicate success or failure. A contract could even
+simply return `null`, as failure is handled separately by aborting. Moreover, the
+contracts we have written so fare always returned the original unmodified
+value upon success, which doesn't sound very useful: after all, the caller and
+the interpreter already had access to this value to begin with.
+
+The motivation for this return value relates to laziness. Nickel is a lazy
```

Comment:
Also function contracts.

Reply:
True, and I thought about bringing that up, but in the end, function contracts are native and their implementation really relevant to the API of user-defined contracts. It turns out we can implement them in user code fully (well that's already half a lie, because they need specific primops for error reporting like `%go_dom%` and `%go_codom%`), but I think the right motivation for the API of user-defined contracts returning a value is to empower users to write lazy contracts. If that's not the case, and it's just for function and built-in contracts, we should arguably have a predicate API for user-defined contracts, and keep the return-value representation as an internal implementation detail.

Reply:
I see your angle.

That being said, there is an argument to make as to whether contracts are like this because Nickel has lazy values, or whether Nickel has lazy values because contracts need it.

Reply:
It's true that the text is implying a stronger causal ordering "lazy evaluation => lazy contracts" when it is rather "we want lazy configurations => lazy evaluation /\ lazy contracts". Will try to tame it down in the next pass.

Reply:
Reworded this part to shift the focus from "lazy evaluation" to "laziness" as a more general property.